packages/aws-cdk/lib/contextplugins.ts

```diff
@@ -48,21 +53,85 @@ export class SSMContextProviderPlugin implements ContextProviderPlugin {
   }
 }
 
+export class HostedZoneContextProviderPlugin implements ContextProviderPlugin {
+
+  constructor(private readonly aws: SDK) {
+  }
+
+  public async getValue(args: {[key: string]: any}) {
+    const account = args.account;
+    const region = args.region;
+    if (!this.isHostedZoneProps(args)) {
+      throw new Error(`HostedZoneProvider requires domainName property to be set in ${args.props}`);
+    }
+    const domainName = args.domainName;
+    debug(`Reading hosted zone ${account}:${region}:${domainName}`);
+    const r53 = await this.aws.route53(account, region, Mode.ForReading);
+    const response = await r53.listHostedZonesByName({ DNSName: domainName }).promise();
+    if (!response.HostedZones) {
+      throw new Error(`Hosted Zone not found in account ${account}, region ${region}: ${domainName}`);
+    }
+    const candidateZones = await this.filterZones(r53, response.HostedZones,
+      !!args.privateZone, args.vpcId);
+    if (candidateZones.length > 1) {
+      const filteProps = `dns:${domainName}, privateZone:${args.privateZone}, vpcId:${args.vpcId}`;
+      throw new Error(`Found more than one matching HostedZone ${candidateZones} for ${filteProps}`);
+    }
+    return candidateZones[0];
+  }
+
+  private async filterZones(
+    r53: AWS.Route53, zones: AWS.Route53.HostedZone[],
+    privateZone: boolean,
+    vpcId: string | undefined): Promise<AWS.Route53.HostedZone[]> {
+
+      let candidates: AWS.Route53.HostedZone[] = [];
+      if (privateZone) {
+        candidates = zones.filter(zone => zone.Config && zone.Config.PrivateZone);
+      } else {
+        candidates = zones.filter(zone => !zone.Config || !zone.Config.PrivateZone);
+      }
+      if (vpcId) {
+        const vpcZones: AWS.Route53.HostedZone[] = [];
+        for (const zone of candidates) {
+          await r53.getHostedZone({Id: zone.Id}, (err, data) => {
+            if (err) {
+              throw new Error(err.message);
```

Comment:
Not that it matters because you're going to be rewriting this call sequyence anyway, but throwing an `Error` in an async callback is almost never what you want to be doing.

In an async callback your call stack is gone: you're basically a couple of stack frames deep in an event loop, and throwing an error is basically only going to abort your callback and the rest of the failure is lost. To pull the thread all the way through, you'll generally call another callback with the error (boooh) or report the error via a promise (yay).

But the `await` syntax takes care of all of that for you.